### PR TITLE
Reject refresh usage in bulk items when using and fix NPE when no source

### LINKED ISSUE #7361: Refresh on specific items in bulk request is ignored when using java api
The Java api can be a bit misleading (https://github.com/elasticsearch/elasticsearch/blob/ba8df3b5bac0eb1317ad49bb9252c29fefc01294/src/main/java/org/elasticsearch/action/update/UpdateRequest.java#L353).  When setting refresh at the individual item level within a bulk request, the refresh does not actually occur:

```java
            UpdateRequest updateRequest = new UpdateRequest()
                    .index("index_name")
                    .type("type")
                    .id("12345")
                    .script("ctx._source.something=param;")
                    .addScriptParam("param", "abcd")
                    .refresh(true);
            BulkRequestBuilder bulkRequest =
                    client.prepareBulk();
            bulkRequest.add(updateRequest);
            bulkRequest.execute().actionGet();
```

The refresh does work if you set it at the bulk request level, eg.
```java
            UpdateRequest updateRequest = new UpdateRequest()
                    .index("index_name")
                    .type("type")
                    .id("12345")
                    .script("ctx._source.something=param;")
                    .addScriptParam("param", "abcde");
            BulkRequestBuilder bulkRequest =
                    client.prepareBulk();
            bulkRequest.add(updateRequest);
            bulkRequest.setRefresh(true).execute().actionGet();
```

### PR DESCRIPTION
**Pull request by dadoonet**
The REST bulk API rejects use of `refresh` at the item level. But the Java API lets the user setting it.

We need to have the same behavior and don't let think the user he can define `refresh` per bulk item.

Note that the user can still define `refresh` on the bulk itself.

Closes #7361.

### CODE REVIEW

**Comment by dadoonet** [comment]
@bleskes Could you review it please?

File: core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 1]
```diff
@@ -125,6 +132,9 @@ public BulkRequest add(IndexRequest request, @Nullable Object payload) {
     }
 
     BulkRequest internalAdd(IndexRequest request, @Nullable Object payload) {
+        if (request.refresh()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: we need to do this in the validate method. The list of requests is exposed and people can change them after adding.

**Comment by bleskes** [comment]
left a minor comment. Thanks for picking this up?

**Comment by dadoonet** [comment]
@bleskes Updated with another commit. BTW I found a NPE in another part of the test while coding tests.

File: core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 2]
```diff
@@ -127,7 +134,7 @@ public BulkRequest add(IndexRequest request, @Nullable Object payload) {
     BulkRequest internalAdd(IndexRequest request, @Nullable Object payload) {
         requests.add(request);
         addPayload(payload);
-        sizeInBytes += request.source().length() + REQUEST_OVERHEAD;
+        sizeInBytes += (request.source() != null ? request.source().length() : 0) + REQUEST_OVERHEAD;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: can we add a note that the lack of source is validated in validate?

File: core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 3]
```diff
@@ -478,8 +485,24 @@ public ActionRequestValidationException validate() {
         if (requests.isEmpty()) {
             validationException = addValidationError("no requests added", validationException);
         }
-        for (int i = 0; i < requests.size(); i++) {
-            ActionRequestValidationException ex = requests.get(i).validate();
+        for (ActionRequest request : requests) {
+            // We first check if refresh has been set
+            if (request instanceof DeleteRequest) {
+                if (((DeleteRequest)request).refresh()) {
+                    validationException = addValidationError("Refresh is not supported for DeleteRequest when added to a BulkRequest", validationException);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: maybe be nice and add that you reuse the bulk request.  So "refresh is not supported on an item request, set the refresh flag on the BulkRequest instead".

**Comment by bleskes** [comment]
LGTM. Left two minor suggestions.

**Comment by dadoonet** [comment]
@bleskes I added another commit and I'd prefer you double check before pushing!

Thanks!

**Comment by bleskes** [comment]
LGTM

**Pull request merged**